### ISSUE TITLE
bug: Using valibot validation creates type errors

### ISSUE THREAD

**Issue by aiakared** [trpc/trpc#4737]
### Provide environment information

  System:
    OS: macOS 13.0.1
    CPU: (8) arm64 Apple M1
    Memory: 59.02 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 20.5.0 - ~/.volta/tools/image/node/20.5.0/bin/node
    Yarn: 4.0.0-rc.43 - ~/.volta/tools/image/yarn/4.0.0-rc.43/bin/yarn
    npm: 9.8.0 - ~/.volta/tools/image/node/20.5.0/bin/npm
    pnpm: 8.6.10 - ~/.volta/bin/pnpm
  Browsers:
    Chrome: 113.0.5672.126
    Edge: 114.0.1823.79
    Firefox Nightly: 89.0a1
    Safari: 16.1
  npmPackages:
    @trpc/client: ^10.37.1 => 10.37.1 
    @trpc/server: ^10.37.1 => 10.37.1

### Describe the bug

Using a simple valibot validator for validation outputs a type error

### Link to reproduction

https://stackblitz.com/edit/js-yma3yp?file=index.ts

### To reproduce

1) Install `@trpc/server@10.37.1` and `valibot@0.13.0`
2) Create a procedure using a valibot input validator
```ts
t.procedure.input(string()).query(({ input }) => `Hello, ${input}!`);
```
<img width="848" alt="image" src="https://github.com/trpc/trpc/assets/70045485/c89c0253-bd80-4d16-8275-9a280fcfea2d">

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by iduuck** [comment]
Seems like this is a versioning bug. With `valibot@^0.8.0` it [is working for me](https://stackblitz.com/edit/js-hms2xy?file=package.json,index.ts).

**Comment by sachinraja** [comment]
see #4715

**Comment by fabian-hiller** [comment]
Unfortunately we could not find a [solution](https://github.com/trpc/trpc/pull/4715) for tRPC with Valibot >=v0.13.0 yet. Currently you can either explicitly set your Valibot version to v0.12.0 or call `parse` on your own.

Downgrade to Valibot <=v0.12.0:

```bash
npm install valibot@0.12.0     # npm
yarn add valibot@0.12.0        # yarn
pnpm add valibot@0.12.0        # pnpm
bun add valibot@0.12.0         # bun
```

Or use Valibot >=v0.13.0 with `parse`:

```ts
.input((input) => parse(YourSchema, input))
```

**Comment by decs** [comment]
You can also wrap it with `typeschema` https://github.com/decs/typeschema#trpc

```ts
import {wrap} from '@decs/typeschema';
import {initTRPC} from '@trpc/server';
import {object, string} from 'valibot';

// Use your favorite validation library, e.g. `valibot`
const schema = object({name: string()});

const t = initTRPC.create();
const appRouter = t.router({
  hello: t.procedure
    .input(wrap(schema)) // like this
    .query(({input}) => `Hello, ${input.name}!`),
});
```

**Comment by malbernaz** [comment]
For anyone arriving in this thread this seems to be the cleanest way to do it:

```ts
import { object, string, parse } from 'valibot';

const schema = object({name: string()});

const appRouter = t.router({
  hello: t.procedure
    .input(i => parse(schema, i))
    .query(({input}) => `Hello, ${input.name}!`),
});
```

**Comment by KATT** [comment]
As mentioned in https://github.com/trpc/trpc/pull/4715#issuecomment-1693109599, we won't support Valibot without using something like @decs' package since Valibot doesn't have a `.parse`-method on the validation objects that are created.

Up for grabs to update the docs using @decs' package

**Comment by fabian-hiller** [comment]
@KATT will update the tRPC docs today or tomorrow.

**Comment by fabian-hiller** [comment]
@decs does `wrap` can also handle the input and output type when a transformation is applied?

**Comment by decs** [comment]
@fabian-hiller it does. it reuses the same api from zod (_output/_input) to expose Infer/InferIn https://github.com/decs/typeschema/blob/main/src/wrap.ts

**Comment by fabian-hiller** [comment]
@decs `wrap` seems not to work when using it in the `.output(…)` validation and transformation of tRPC.

**Comment by fabian-hiller** [comment]
The problem is that tRPC uses `.assert` from `wrap` and therefore with the current implementation of tRPC it ignores the output of the schema: https://github.com/trpc/trpc/blob/main/packages/server/src/core/internals/getParseFn.ts#L34-L39

**Comment by fabian-hiller** [comment]
A quick fix could be to add `.parse` to `wrap`.

**Comment by decs** [comment]
Good catch! added the `parse` function on `v0.11.0` and improved our integration tests on tRPC to cover this case
